cashu/wallet/wallet.py

```diff
@@ -539,29 +604,259 @@ def _meltrequest_include_fields(proofs):
+        cls,
+        url: str,
+        db: str,
+        name: str = "no_name",
+        skip_private_key: bool = False,
+    ):
+        """Initializes a wallet with a database and initializes the private key.
+
+        Args:
+            url (str): URL of the mint.
+            db (str): Path to the database.
+            name (str, optional): Name of the wallet. Defaults to "no_name".
+            skip_private_key (bool, optional): If true, the private key is not initialized. Defaults to False.
+
+        Returns:
+            Wallet: Initialized wallet.
+        """
+        self = cls(url=url, db=db, name=name)
+        await self._migrate_database()
+        if not skip_private_key:
+            await self._init_private_key()
+        return self
+
+    async def _migrate_database(self):
+        try:
+            await migrate_databases(self.db, migrations)
+        except Exception as e:
+            logger.error(f"Could not run migrations: {e}")
+
+    async def _init_private_key(self, from_mnemonic: Optional[str] = None) -> None:
+        """Initializes the private key of the wallet from the mnemonic.
+        There are three ways to initialize the private key:
+        1. If the database does not contain a seed, and no mnemonic is given, a new seed is generated.
+        2. If the database does not contain a seed, and a mnemonic is given, the seed is generated from the mnemonic.
+        3. If the database contains a seed, the seed is loaded from the database.
+
+        If the mnemonic was not loaded from the database, the seed and mnemonic are stored in the database.
+
+        Args:
+            from_mnemonic (Optional[str], optional): Mnemonic to use. Defaults to None.
+
+        Raises:
+            ValueError: If the mnemonic is not BIP39 compliant.
+        """
+        ret_db = await get_seed_and_mnemonic(self.db)
+
+        mnemo = Mnemonic("english")
+
+        if ret_db is None and from_mnemonic is None:
+            # if there is no seed in the database, generate a new one
+            mnemonic_str = mnemo.generate()
+            wallet_command_prefix_str = (
+                f" --wallet {settings.wallet_name}"
+                if settings.wallet_name != "wallet"
+                else ""
+            )
+            wallet_name = (
+                f' for wallet "{settings.wallet_name}"'
+                if settings.wallet_name != "wallet"
+                else ""
+            )
+            print(
+                f'Generated a new mnemonic{wallet_name}. To view it, run "cashu{wallet_command_prefix_str} info --mnemonic".'
+            )
+        elif from_mnemonic:
+            # or use the one provided
+            mnemonic_str = from_mnemonic
+        elif ret_db is not None:
+            # if there is a seed in the database, use it
+            _, mnemonic_str = ret_db[0], ret_db[1]
+        else:
+            logger.debug("No mnemonic provided")
+            return
+
+        if not mnemo.check(mnemonic_str):

```

Comment:
Consider adding .lower() to the seed when it is used so that the re-entry of the seed isn't case sensitive. No need for it to be. Tweaked my local copy and it was working nicely without throwing the "Invalid mnemonic" for mixed casing seeds.

Also maybe a bit more complex but could be useful to allow seed restoration with just first 4 (or 3) characters as no need for the user to write them all out. Maybe this is a nice-to-have though, if it is a lot of work maybe not worth it.